Parse/Infrastructure/CacheController.cs

```diff
@@ -10,253 +10,304 @@
 using Parse.Infrastructure.Utilities;
 using static Parse.Resources;
 
-namespace Parse.Infrastructure
+namespace Parse.Infrastructure;
+
+/// <summary>
+/// Implements `IStorageController` for PCL targets, based off of PCLStorage.
+/// </summary>
+public class CacheController : IDiskFileCacheController
 {
-    /// <summary>
-    /// Implements `IStorageController` for PCL targets, based off of PCLStorage.
-    /// </summary>
-    public class CacheController : IDiskFileCacheController
+    class FileBackedCache : IDataCache<string, object>
     {
-        class FileBackedCache : IDataCache<string, object>
-        {
-            public FileBackedCache(FileInfo file) => File = file;
+        public FileBackedCache(FileInfo file) => File = file;
 
-            internal Task SaveAsync() => Lock(() => File.WriteContentAsync(JsonUtilities.Encode(Storage)));
+        internal async Task SaveAsync()
+        {
+            await LockAsync(() => File.WriteContentAsync(JsonUtilities.Encode(Storage)));
+        }
 
-            internal Task LoadAsync() => File.ReadAllTextAsync().ContinueWith(task =>
+        internal async Task LoadAsync()
+        {
+            try
             {
+                string fileContent = await File.ReadAllTextAsync();
                 lock (Mutex)
                 {
-                    try
-                    {
-                        Storage = JsonUtilities.Parse(task.Result) as Dictionary<string, object>;
-                    }
-                    catch
-                    {
-                        Storage = new Dictionary<string, object> { };
-                    }
+                    Storage = JsonUtilities.Parse(fileContent) as Dictionary<string, object> ?? new Dictionary<string, object>();
                 }
-            });
-
-            // TODO: Check if the call to ToDictionary is necessary here considering contents is IDictionary<string object>.
-
-            internal void Update(IDictionary<string, object> contents) => Lock(() => Storage = contents.ToDictionary(element => element.Key, element => element.Value));
-
-            public Task AddAsync(string key, object value)
+            }
+            catch
             {
                 lock (Mutex)
                 {
-                    Storage[key] = value;
-                    return SaveAsync();
+                    Storage = new Dictionary<string, object>();
                 }
             }
+        }
+
+        internal void Update(IDictionary<string, object> contents)
+        {
+            Lock(() => Storage = contents.ToDictionary(element => element.Key, element => element.Value));
+        }
+
+        public async Task AddAsync(string key, object value)
+        {
+            lock (Mutex)
+            {
+                Storage[key] = value;
+            }
+            await SaveAsync();
```

Comment:
Previous implementation blocks concurrent threads from modifying Storage and initiating SaveAsync() until SaveAsync() returns the Task. This can cause potential deadlocks if SaveAsync() depends on the same lock.

New implementation only blocks concurrent threads while updating Storage. SaveAsync() is executed outside the lock, avoiding deadlocks but risking race conditions if SaveAsync() accesses Storage.

Is this behavior change intended?

Would suggest maybe implementing a semaphore for better asynchronous synchronization.

Reply:
On this now